mozjs-sys/build.rs

```diff
@@ -57,17 +57,42 @@ fn main() {
     // Used by mozjs downstream, don't remove.
     println!("cargo:outdir={}", build_dir.display());
 
-    if let Some(archive) = env::var_os("MOZJS_ARCHIVE") {
-        download_static_lib_binaries(&PathBuf::from(archive), &build_dir);
+    // Link to pre-built archive first if it exists.
+    let create_archive = env::var_os("MOZJS_CREATE_ARCHIVE").is_some();
+    let build_from_source = if env::var_os("MOZJS_FROM_SOURCE").is_some() {
+        println!("Environment variable MOZJS_FROM_SOURCE is set. Building from source directly.");
+        true
+    } else if create_archive {
+        println!(
+            "Environment variable MOZJS_CREATE_ARCHIVE is set. Building from source directly."
+        );
+        true
+    } else if env::var_os("CARGO_FEATURE_DEBUGMOZJS").is_some() {
+        println!("debug-mozjs feature is enabled. Building from source directly.");
+        true
+    } else if !env::var_os("CARGO_FEATURE_STREAMS").is_some() {
+        println!("streams feature isn't enabled. Building from source directly.");
+        true
     } else {
+        match link_static_lib_binaries(&build_dir) {
+            Ok(()) => false,
+            Err(e) => {
+                println!("cargo:warning=Failed to link pre-built archive by {e}. Building from source instead.");
+                true
+            }
+        }
+    };
+
+    // Builing from source if there's no archive.
+    if build_from_source {
         fs::create_dir_all(&build_dir).expect("could not create build dir");
         build_spidermonkey(&build_dir);
         build_jsapi(&build_dir);
         build_jsapi_bindings(&build_dir);
         jsglue::build(&build_dir);
 
         // If this env variable is set, create the compressed tarball of spidermonkey.
-        if env::var_os("MOZJS_CREATE_ARCHIVE").is_some() {
+        if create_archive {
             compress_static_lib(&build_dir).expect("Failed to compress static lib binaries.");
         }
     }
```

Comment:
These two parts of the code are alternatives, so they should be two bodies of an if-else statement.